bin/knc.c

```diff
@@ -1072,42 +1111,50 @@ move_data(work_t *work)
 		if (reap() > 0)
 			LOG(LOG_NOTICE, ("child died before EOF"));
 
-		FD_ZERO(&rdset);
-		FD_ZERO(&wrset);
-
-		/* Add stderr */
-		if (work->local_err != -1)
-			FD_SET(work->local_err, &rdset);
+		pfd_net->events = POLLRDHUP;
+		pfd_in->events  = POLLRDHUP;
+		pfd_out->events = 0;
+		pfd_err->events = POLLRDHUP;
 
 		/*
 		 * We have this timeout only to allow us to recover
```

Comment:
Ah, so maybe we could have a signal handler for `SIGCHLD` to write a single byte to a self-pipe we could add the read end of to the event loop, then we wouldn't need the timeout at all.

Reply:
The code already has a handler for SIGCHLD and ignores it.  The handler also does not set SA_RESTART.

Given that the program has at most one child that it creates, we could just set a volatile sig_atomic_t static var upon receiving SIGCHLD and check for that, not needing a self-pipe at all.

Reply:
If child dies, I think we should waitpid() in the signal handler for SIGCHLD and set a flag.  In move_data(), if that flag is set, then we should read and discard data from the network, and as soon as local buffers (read from child and sent to network) are flushed, then exit.